src/metabase/pulse/attachment.clj

```diff
@@ -0,0 +1,97 @@
+(ns metabase.pulse.attachment
+  (:require
+   [clojure.java.io :as io]
+   [java-time.api :as t]
+   [metabase.driver :as driver]
+   [metabase.driver.util :as driver.u]
+   [metabase.query-processor.store :as qp.store]
+   [metabase.query-processor.streaming :as qp.streaming]
+   [metabase.query-processor.streaming.interface :as qp.si]
+   [metabase.query-processor.streaming.xlsx :as qp.xlsx]
+   [metabase.util :as u]
+   [metabase.util.date-2 :as u.date]
+   [metabase.util.i18n :as i18n :refer [tru]])
+  (:import
+   (java.io File IOException OutputStream)))
+
+(set! *warn-on-reflection* true)
+
+(defn- create-temp-file
+  "Separate from `create-temp-file-or-throw` primarily so that we can simulate exceptions in tests"
+  [filename suffix]
+  (doto (File/createTempFile filename (str "." suffix))
+    .deleteOnExit))
+
+(defn create-temp-file-or-throw
+  "Tries to create a temp file, will give the users a better error message if we are unable to create the temp file"
+  [filename suffix]
+  (try
+    (create-temp-file filename suffix)
+    (catch IOException e
+      (let [ex-msg (tru "Unable to create temp file in `{0}` for email attachments "
+                        (System/getProperty "java.io.tmpdir"))]
+        (throw (IOException. ex-msg e))))))
+
+(defn text-file-writer
+  "Writes given text to a file via the given output-stream."
+  [output-stream ^String text]
+  (let [writer (io/writer output-stream)]
+    (.write writer text)
+    (.flush writer)))
+
+(defn- stream-api-results-to-export-format
+  "For legacy compatibility. Takes QP results in the normal `:api` response format and streams them to a different
+  format.
+
+  TODO -- this function is provided mainly because rewriting all the Pulse/Alert code to stream results directly
```

Comment:
This TODO wasn't my own. 

It should probably be addressed in a tech-debt issue/PR.